manual/src/cmds/runtime-tracing.etex

```diff
@@ -314,4 +322,146 @@ Percentile 	 Latency (ms)
+  inside a \texttt{Runtime_events.Callbacks.t} object to be used when polling
+  (see section~\ref{s:runtime-tracing-ocaml-apis} above).
+\item \emph{(Optional, if there are distinct events registered with the same
+  custom type:)} In that callback, pattern-match on the event tag in order to
+  discriminate between different event tags of the same custom type.
+\end{enumerate}
+
+Note that if the emitter and the consumer are different programs, both must
+perform steps 1 to 3 to register custom events and custom event types (if any).
+
+Unregistered events for user-defined types are currently ignored. As a fallback,
+unregistered, custom events of a built-in type are available, but are all
+tagged \texttt{UNK} (unknown).
+
+There are thus three cases for the consumer when an event is received:
+\begin{itemize}
+\item event is registered: payload and tag are available.
+\item event is not registered and has a built-in event type (unit, int, span):
+    only the payload is available.
+\item event is not registered and has a custom event type: event is dropped.
+\end{itemize}
+
+Note that event consumers bind callbacks to event \emph{types}, so they can work
+as generic listeners interpreting payloads coming from events that were not
+registered. Because this only works for events of built-in types, it can be
+useful to emit events in pairs: an event of a custom type with an event of a
+built-in type, enabling the design of a specialized consumer while staying
+compatible with generic tracing tools.
+
+\subsection{s:runtime-tracing-custom-events-usage}{Tracing custom events: an
+example}
+
+The following program uses the \texttt{Runtime_events.User} and
+\texttt{Runtime_events.Type} modules to declare two custom events providing
+\texttt{span} and \texttt{int} values. The \texttt{tag} extensible variant is
+extended with \texttt{CustomSpan} and \texttt{CustomInt}.
+
+\begin{verbatim}
+type Runtime_events.User.tag += CustomSpan | CustomInt
+
+let count_span =
+    Runtime_events.User.register "count.span" CustomSpan
+        Runtime_events.Type.span
+
+let count_value =
+    Runtime_events.User.register "count.value" CustomInt
+        Runtime_events.Type.int
+
+let count () =
+    Runtime_events.User.write count_span Begin;
+    for i = 1 to 5 do
+        Runtime_events.User.write count_value i
+    done;
+    Runtime_events.User.write count_span End
+
+let () =
+    Runtime_events.start ();
+    for _ = 1 to 3 do
+        count ()
+    done
+\end{verbatim}
+
+On the consumer side, one can use the provided event tag and type to match on
+the relevant events.
+
+\begin{verbatim}
+let span_event_handler domain_id ts event value =
+    (* we're only interested in our CustomSpan event *)
+    match Runtime_events.User.tag event, value with
+    | CustomSpan, Runtime_events.Type.Begin -> Printf.printf "> count begin\n"
+    | CustomSpan, End -> Printf.printf "< count end\n"
+    | _ -> ()
```

Comment:
Is this "consumer side" example supposed to be a self-contained program? Where is `CustomSpan` defined? Are we actually linking with the "producer side" program above to share the same `CustomSpan` definition? (That would be a weird idea given that the producer program performs side-effects.) I am confused.

Reply:
Okay, for maximal completeness, I duplicated the new constructors definitions as well as the registering, and added a sentence to explain it. fd83647e5a

Reply:
It is not just for completeness: demonstrating that it is okay to register (tags and) events separately is clearer than just saying so somewhere. The previous approach was confusing because if someone asked themselves the question, they could notice that the code example were written specifically to *not* answer the question of whether the definitions (in the example) are shared or duplicated.

Reply:
Yes, I agree. Completeness wasn’t the right word.